sui_programmability/examples/nfts/sources/marketplace.move

```diff
@@ -1,102 +1,168 @@
+    /// price in [`Coin<COIN>`].
+    struct Listing has key, store {
+        id: UID,
+        ask: u64,
         owner: address,
     }
 
     /// Create a new shared Marketplace.
-    public entry fun create(ctx: &mut TxContext) {
+    public entry fun create<COIN>(ctx: &mut TxContext) {
         let id = object::new(ctx);
-        let marketplace = Marketplace { id };
-        transfer::share_object(marketplace);
+        transfer::share_object(Marketplace<COIN> { id })
     }
 
     /// List an item at the Marketplace.
-    public entry fun list<T: key + store, C>(
-        marketplace: &mut Marketplace,
+    public entry fun list<T: key + store, COIN>(
+        marketplace: &mut Marketplace<COIN>,
         item: T,
         ask: u64,
         ctx: &mut TxContext
     ) {
         let item_id = object::id(&item);
-        let listing = Listing<T, C> {
-            item,
+        let listing = Listing {
             ask,
+            id: object::new(ctx),
             owner: tx_context::sender(ctx),
         };
-        dynamic_field::add(&mut marketplace.id, item_id, listing);
+
+        ofield::add(&mut listing.id, true, item);
+        ofield::add(&mut marketplace.id, item_id, listing)
     }
 
     /// Remove listing and get an item back. Only owner can do that.
-    public fun delist<T: key + store, C>(
-        marketplace: &mut Marketplace,
+    public fun delist<T: key + store, COIN>(
+        marketplace: &mut Marketplace<COIN>,
         item_id: ID,
         ctx: &mut TxContext
     ): T {
-        let Listing<T, C> { item, ask: _, owner } =
-            dynamic_field::remove(&mut marketplace.id, item_id);
+        let Listing {
+            id,
+            owner,
+            ask: _,
+        } = ofield::remove(&mut marketplace.id, item_id);
 
         assert!(tx_context::sender(ctx) == owner, ENotOwner);
 
+        let item = ofield::remove(&mut id, true);
+        object::delete(id);
         item
     }
 
     /// Call [`delist`] and transfer item to the sender.
-    public entry fun delist_and_take<T: key + store, C>(
-        marketplace: &mut Marketplace,
+    public entry fun delist_and_take<T: key + store, COIN>(
+        marketplace: &mut Marketplace<COIN>,
         item_id: ID,
         ctx: &mut TxContext
     ) {
-        let item = delist<T, C>(marketplace, item_id, ctx);
+        let item = delist<T, COIN>(marketplace, item_id, ctx);
         transfer::transfer(item, tx_context::sender(ctx));
     }
 
     /// Purchase an item using a known Listing. Payment is done in Coin<C>.
     /// Amount paid must match the requested amount. If conditions are met,
     /// owner of the item gets the payment and buyer receives their item.
-    public fun buy<T: key + store, C>(
-        marketplace: &mut Marketplace,
+    public fun buy<T: key + store, COIN>(
+        marketplace: &mut Marketplace<COIN>,
         item_id: ID,
-        paid: Coin<C>,
+        paid: Coin<COIN>,
```

Comment:
I wouldn't mind keeping this simple by just hardcoding `Coin<SUI>` as the accepted form of payment--should make this example code easier to understand.